app/controllers/api/base_controller/authentication.rb

```diff
@@ -12,22 +12,25 @@ def require_api_user_or_token
         elsif request.headers[HttpHeaders::AUTH_TOKEN]
           authenticate_with_user_token(request.headers[HttpHeaders::AUTH_TOKEN])
         else
-          authenticate_options = {
-            :require_user => true,
-            :timeout      => ::Settings.api.authentication_timeout.to_i_with_method
-          }
-
-          if (user = authenticate_with_http_basic { |u, p| User.authenticate(u, p, request, authenticate_options) })
-            auth_user_obj = userid_to_userobj(user.userid)
-            authorize_user_group(auth_user_obj)
-            validate_user_identity(auth_user_obj)
-            User.current_user = auth_user_obj
-          else
-            request_http_basic_authentication
-            return
+          success = authenticate_with_http_basic do |u, p|
+            begin
+              timeout = ::Settings.api.authentication_timeout.to_i_with_method
+              user = User.authenticate(u, p, request, :require_user => true, :timeout => timeout)
+              auth_user_obj = userid_to_userobj(user.userid)
+              authorize_user_group(auth_user_obj)
+              validate_user_identity(auth_user_obj)
+              User.current_user = auth_user_obj
+            rescue MiqException::MiqEVMLoginError => e
+              raise AuthenticationError, e.message
+            end
           end
+          raise AuthenticationError unless success
         end
         log_api_auth
+      rescue AuthenticationError => e
+        api_log_error("AuthenticationError: #{e.message}")
+        response.headers["Content-Type"] = "application/json"
+        request_http_basic_authentication("Application", ErrorSerializer.new(:unauthorized, e).serialize.to_json)
```

Comment:
We're missing the logging of the 401,

before this PR, we get the following:

```
[----] I, [2018-04-16T17:48:16.502716 #26595:3fdeb52eb958]  INFO -- : MIQ(Api::ApiController.log_request_initiated)  
[----] I, [2018-04-16T17:48:16.502900 #26595:3fdeb52eb958]  INFO -- : MIQ(Api::ApiController.log_request) API Request:    {:requested_at=>"2018-04-16 21:48:16 UTC", :method=>"GET", :url=>"http://localhost:3000/api"}
[----] E, [2018-04-16T17:48:16.571327 #26595:3fdeb52eb958] ERROR -- : MIQ(Api::ApiController.api_error) API Error
[----] E, [2018-04-16T17:48:16.571381 #26595:3fdeb52eb958] ERROR -- : MIQ(Api::ApiController.api_error) MiqException::MiqEVMLoginError: Authentication failed
[----] I, [2018-04-16T17:48:16.571986 #26595:3fdeb52eb958]  INFO -- : MIQ(Api::ApiController.log_request) Response:       {:completed_at=>"2018-04-16 21:48:16 UTC", :size=>"0.108 KBytes", :time_taken=>"0.069 Seconds", :status=>401}
```

The last Response: entry is missing.  This stuff is handled by api_error, so instead of adding more of that logic here, maybe the above can be as follows:

```
rescue AuthenticationError => e
   api_error(:unauthorized, e)
   request_http_basic_authentication("Application", ErrorSerializer.new(:unauthorized, e).serialize.to_json)
end
```

Or possibly get api_error to return to the response generated

```
rescue AuthenticationError => e
   request_http_basic_authentication("Application", api_error(:unauthorized, e).to_json)
end
```

Reply:
Interesting. It looks like `request_http_basic_authentication` is short-circuiting the `after_action` that invokes this - it's not specific to `api_error`. Will look into it.

Reply:
Addressed in https://github.com/ManageIQ/manageiq-api/pull/359/commits/9fe4247787aaa858a1cb9ddfc2413f8ca12318ee

Reply:
Thanks @imtayadeway for the update.